### DISCUSSION TITLE
Speeding up topics_over_time

### DISCUSSION THREAD

**Discussion by samuelwillis1** [MaartenGr/BERTopic#1581] [Q&A]
Hi! I was wondering if anybody had any experience in speeding up the `topics_over_time` method, as it seems to be the bottleneck currently for runs of the pipeline I'm working on.

In general, it can easily take 2-3 times longer to run the `topics_over_time` step than it does to run the `fit_transform` step. This is being done on Kubernetes where the resources provided are only cpu and ram. We know about the impacts of running on cpu during the modelling step without making changes to sentence transformers etc, is this something that affects the `topics_over_time` method too?

Thanks in advance :)

**Comment by MaartenGr** [comment]
The `topics_over_time` method is essentially running the c-TF-IDF calculation as many times as there are unique timestamps. Increasing the number of unique timestamps (or bins) will also increase the computation time. 

Having said that, could you share your code? Without it, it is not possible to make specific statements about your situation.

**Reply by samuelwillis1** [reply]
Hi Maarten, thanks for getting back so quickly!

Below is the code I'm using concerning the `topics_over_time` method, where `self.model` is the BERTopic model already trained on my data, and `self.text`/`self.timestamps` are as expected. This is on about 21k documents over a 6 month period, so the number of bins is to try and roughly split the data by month.

```
       topics_over_time = self.model.topics_over_time(
            docs=self.text, timestamps=self.timestamps, datetime_format="%Y-%m-%d", nr_bins=6
        )
        topics_over_time["dynamic_label_date"] = pd.to_datetime(topics_over_time["Timestamp"], format="%Y-%m-%d")
        topics_over_time = topics_over_time.rename(columns={"Topic": "topic_id", "Words": "dynamic_label"})

        # Merge data with topics over time dataframe
        data_topics = data.merge(
            topics_over_time[["topic_id", "dynamic_label", "dynamic_label_date"]], on="topic_id", how="left"
        )
```

For more context, when I run the code locally from my machine on the same amount of documents it is generally quite fast (~a couple of minutes, quicker than the modelling step at least). Whereas when running on Kubernetes, this is the step that seems to increase in time past what I would expect (seen it take about an hour previously). I'm starting to assume it may be a resource problem, but have increased RAM and CPU availability quite significantly and it doesn't seem to help much.

Open to any suggestions! And thanks again for taking the time to have a look!

**Reply by samuelwillis1** [reply]
Apologies for all the edits, they were all for formatting issues I was encountering!

**Reply by MaartenGr** [reply]
Could you also share how you initialized BERTopic? That will also give information on how the representation model(s) were formed. Moreover, after training, how many topics were created?

**Reply by samuelwillis1** [reply]
Sorry for the quite long code snippet below, just trying to include it all in one spot for you to see:

```
stop_words = text.ENGLISH_STOP_WORDS.union(stops_lower)
vectorizer_model = text.CountVectorizer(stop_words=stop_words, ngram_range=(1, 3))

umap_model = UMAP(random_state=umap_params["random_state"], n_neighbors=15)

hdbscan_model = HDBSCAN(
            cluster_selection_epsilon=0.075,
            min_cluster_size=8,
            min_samples=5,
        )

# Create vectorizer model, clustering model, and dimensionality-reduction model
        logger.info("Building vectorizer, HDBSCAN, and UMAP")
        representation_model = MaximalMarginalRelevance(diversity=0.5)
        vectorizer_model = vectorizer_model
        hdbscan_model = hdbscan_model
        umap_model = umap_model

        model = BERTopic(
            vectorizer_model=vectorizer_model,
            hdbscan_model=hdbscan_model,
            umap_model=umap_model,
            representation_model=representation_model,
        )
```

Before tuning parameters there were about 630 topics, afterwards there are around 350 though the `topics_over_time` still takes significantly longer. Thanks again!

**Reply by MaartenGr** [reply]
> Before tuning parameters there were about 630 topics, afterwards there are around 350 though the topics_over_time still takes significantly longer. Thanks again!

Do note that you cannot compare the wall time between `fit_transform` and `topics_over_time` one on one since they are quite different procedures. One is not supposed to be faster or slower than the other.

Having said that, you can speed up `topics_over_time` by setting `min_df` to a higher value in the `CountVectorizer`. Most likely, you have generated a large vocabulary with your vectorizer since you have set the ngram range that high. Setting `min_df` to at least 2 is likely to speed up `topics_over_time` since it decreases its vocabulary which affects the c-TF-IDF calculation.

**Reply by samuelwillis1** [reply]
Hi Maarten, thanks for this! It definitely seems like raising `min_df` does speed up the `topics_over_time` as I've noted in run times recently. 

One other thing worth mentioning is that I've tested runtimes both using `MaximalMarginalRelevance` as the representation model as above, and with no representation model set at all. I've noticed a massive decrease in time taken for `topics_over_time` when omitting the MMR as the representation model. For example, one quick test earlier showed it took ~2.5 minutes to compute the topics over time with the representation model set as MMR, and just 6 seconds when omitting it. 

Thank you for your support and suggestion, it will be very helpful with speeding up runtimes!
